shell/browser/api/electron_api_web_request.cc

```diff
@@ -211,41 +214,82 @@ CalculateOnBeforeSendHeadersDelta(const net::HttpRequestHeaders* old_headers,
 gin::WrapperInfo WebRequest::kWrapperInfo = {gin::kEmbedderNativeGin};
 
 WebRequest::RequestFilter::RequestFilter(
-    std::set<URLPattern> url_patterns,
+    std::set<URLPattern> include_url_patterns,
+    std::set<URLPattern> exclude_url_patterns,
     std::set<extensions::WebRequestResourceType> types)
-    : url_patterns_(std::move(url_patterns)), types_(std::move(types)) {}
+    : include_url_patterns_(std::move(include_url_patterns)),
+      exclude_url_patterns_(std::move(exclude_url_patterns)),
+      types_(std::move(types)),
+      is_filter_defined_(false) {}
 WebRequest::RequestFilter::RequestFilter(const RequestFilter&) = default;
 WebRequest::RequestFilter::RequestFilter() = default;
 WebRequest::RequestFilter::~RequestFilter() = default;
 
-void WebRequest::RequestFilter::AddUrlPattern(URLPattern pattern) {
-  url_patterns_.emplace(std::move(pattern));
+void WebRequest::RequestFilter::AddUrlPattern(URLPattern pattern,
+                                              bool is_match_pattern) {
+  if (is_match_pattern) {
+    include_url_patterns_.emplace(std::move(pattern));
+  } else {
+    exclude_url_patterns_.emplace(std::move(pattern));
+  }
 }
 
 void WebRequest::RequestFilter::AddType(
     extensions::WebRequestResourceType type) {
   types_.insert(type);
 }
 
-bool WebRequest::RequestFilter::MatchesURL(const GURL& url) const {
-  if (url_patterns_.empty())
-    return true;
+bool WebRequest::RequestFilter::MatchesURL(
+    const GURL& url,
+    const std::set<URLPattern>& patterns) const {
+  if (patterns.empty())
+    return false;
 
-  for (const auto& pattern : url_patterns_) {
+  for (const auto& pattern : patterns) {
     if (pattern.MatchesURL(url))
       return true;
   }
   return false;
 }
 
+void WebRequest::RequestFilter::SetFilterDefined(bool is_defined) {
+  is_filter_defined_ = is_defined;
+}
+
 bool WebRequest::RequestFilter::MatchesType(
     extensions::WebRequestResourceType type) const {
   return types_.empty() || base::Contains(types_, type);
```

Comment:
Ideally,  same logic as url, `types_.empty()` should mean that  no types are matched.

Reply:
this could be a breaking change if I just remove `types_.empty()` == matched but I don't imagine many ppl using types (which is an optional param) and leaving it as an empty array. 

let me know if anyone has objections to me changing to just below without deprecating this field:

```
bool WebRequest::RequestFilter::MatchesType(
    extensions::WebRequestResourceType type) const {
  return base::Contains(types_, type);
```